File: frontend/src/types/assessment.ts
/** Scope types for quizzes and exam sets. */
export type QuizScopeType = 'knowledge_point' | 'section' | 'chapter' | 'stage' | 'path';
export type ExamScopeType = 'chapter' | 'stage' | 'path';

export type AttemptStatus = 'started' | 'in_progress' | 'submitted' | 'graded' | 'processing' | 'completed' | 'failed' | 'cancelled';
export type ExamSetStatus = 'not_started' | 'in_progress' | 'completed';

/** A question reference inside a quiz or exam set. */
export interface QuestionRef {
  questionId: string;
  type: string;
  stemAbbr?: string;
  difficulty?: string;
  score?: number;
}

/** Lightweight instant quiz — generated on the fly, not archived. */
export interface Quiz {
  id: string;
  title: string;
  sessionId: string;
  scopeType: QuizScopeType;
  scopeId?: string | null;
  pathId?: string | null;
  stageId?: string | null;
  chapterId?: string | null;
  sectionId?: string | null;
  knowledgePointIds?: string[];
  difficulty: string;
  questionCount: number;
  questions?: QuestionRef[];
  linkedQuestions?: LinkedQuestion[];
  source: string;
  archivePolicy: string;
  createdAt: string | null;
}

/** A question resolved from PracticeQuestionModel, linked to a quiz/exam set. */
export interface LinkedQuestion {
  questionId: string;
  type: string;
  stem: string;
  options?: string[] | null;
  difficulty: string;
  knowledgePoints?: string[];
}

/** Heavyweight exam set — persistent, archivable, multi-attempt. */
export interface ExamSet {
  id: string;
  title: string;
  sessionId: string;
  scopeType: ExamScopeType;
  scopeId?: string | null;
  pathId?: string | null;
  stageId?: string | null;
  chapterId?: string | null;
  knowledgePointIds?: string[];
  difficulty: string;
  difficultyDistribution?: Record<string, number> | null;
  questionCount: number;
  questions?: QuestionRef[];
  linkedQuestions?: LinkedQuestion[];
  estimatedMinutes: number;
  totalScore: number;
  status: ExamSetStatus;
  source: string;
  archivePolicy: string;
  createdAt: string | null;
  updatedAt: string | null;
}

/** A single answer record linked to an attempt. */
export interface AttemptAnswer {
  id: number;
  questionId: string;
  studentAnswer: string;
  totalScore: number | null;
  errorType: string | null;
  errorLabel: string | null;
  errorExplanation: string | null;
  suggestions: string[];
  createdAt: string | null;
}

/** One attempt at a quiz or exam set, grouping multiple answer records. */
export interface Attempt {
  id: number;
  attemptId: string;
  sessionId: string;
  subjectId?: string | null;
  quizId?: string | null;
  examSetId?: string | null;
  learnerId?: string | null;
  answers?: AttemptAnswer[];
  linkedAnswers?: AttemptAnswer[];
  totalScore?: number | null;
  maxScore: number;
  status: AttemptStatus;
  attemptNumber?: number;
  idempotencyKey?: string;
  assessmentEligible?: boolean;
  startedAt: string | null;
  submittedAt: string | null;
  gradedAt?: string | null;
  answersRevealedAt?: string | null;
  processingTaskId?: string | null;
  diagnosisTaskId?: string | null;
  createdAt: string | null;
}

/** Per-question grading result after quiz submission. */
export interface QuizResult {
  questionId: string;
  studentAnswer: string;
  isCorrect: boolean;
  score: number;
  maxScore: number;
  correctAnswer?: string;
  explanation?: string;
  feedback?: string;
  errorType?: string | null;
  errorLabel?: string | null;
  knowledgePoint?: string;
}

/** Request body for section quiz generation. */
export interface SectionQuizGenerateRequest {
  sessionId: string;
  title: string;
  knowledgePoints: string[];
  lectureSummary?: string;
  difficulty?: string;
  requirements?: string;
  pathId?: string;
  stageId?: string;
  chapterId?: string;
  sectionId?: string;
}

/** Request body for quiz submission. */
export interface QuizSubmitRequest {
  sessionId: string;
  answers: { questionId: string; answer: string }[];
  /** Client-generated unique key for idempotent submission. Required. */
  idempotencyKey: string;
  /** Whether the learner viewed correct answers before submitting. */
  answersRevealed?: boolean;
  /** Client-side timestamp of submission intent. */
  clientSubmittedAt?: string;
}

/** Response from quiz submission. */
export interface QuizSubmitResponse {
  attempt: Attempt;
  results: QuizResult[];
  totalScore: number;
  maxScore: number;
  sectionStatusSuggestion: 'mastered' | 'in_progress' | 'needs_review';
  weakPoints?: WeakPoint[];
  /** True when this response is a replay of a previously-submitted attempt. */
  idempotentReplay?: boolean;
  /** Per-knowledge-point results, one per mapping per question. */
  knowledgePointResults?: KnowledgePointResult[];
  /** Workflow task ID for post-submit assessment processing (SSE-pollable). */
  processingTaskId?: string | null;
}

/** Per-knowledge-point result computed from a graded answer. */
export interface KnowledgePointResult {
  knowledgePointKey: string;
  knowledgePointLabel: string;
  questionId: string;
  attemptId: string;
  rawScore: number;
  maxScore: number;
  normalizedScore: number;
  weight: number;
  weightedScore: number;
  mappingConfidence: number;
  gradingConfidence: number;
  isCorrect: boolean;
  errorType: string | null;
  assessmentEligible: boolean;
}

/** A knowledge-point-level weakness summary. */
export interface WeakPoint {
  name: string;
  errorCount: number;
  totalAttempts: number;
  errorRate: number;
  latestError?: string;
  errorTypes?: string[];
  masteryEstimate?: number;
  suggestedAction?: string;
  source?: string;
}

/** Request body for exam set generation. */
export interface ExamSetGenerateRequest {
  sessionId: string;
  title: string;
  scopeType: 'chapter' | 'stage' | 'path';
  scopeId?: string;
  pathId?: string;
  stageId?: string;
  chapterId?: string;
  knowledgePointIds?: string[];
  knowledgePoints?: string[];
  difficulty?: string;
  questionCount?: number;
}

/** Aggregated weakness summary response. */
export interface WeaknessSummary {
  weakPoints: WeakPoint[];
  total: number;
}

File: backend/app/schemas/assessment.py
"""Pydantic schemas for quiz, exam-set and attempt data models.

These schemas power the assessment REST API and serve as the
type contract between backend routers and the frontend client.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


# ═══════════════════════════════════════════════════════════════════════
# Shared / enum types
# ═══════════════════════════════════════════════════════════════════════

SCOPE_TYPES_QUIZ = {"knowledge_point", "section", "chapter", "stage", "path"}
SCOPE_TYPES_EXAM = {"chapter", "stage", "path"}
ATTEMPT_STATUSES = {"in_progress", "submitted", "graded"}
EXAM_SET_STATUSES = {"not_started", "in_progress", "completed"}


# ═══════════════════════════════════════════════════════════════════════
# Quiz
# ═══════════════════════════════════════════════════════════════════════


class QuizCreate(BaseModel):
    """Payload for creating a new instant quiz."""

    session_id: str = Field(default="", alias="sessionId")
    title: str = ""
    scope_type: str = Field(default="knowledge_point", alias="scopeType")
    scope_id: str | None = Field(default=None, alias="scopeId")
    path_id: str | None = Field(default=None, alias="pathId")
    stage_id: str | None = Field(default=None, alias="stageId")
    chapter_id: str | None = Field(default=None, alias="chapterId")
    section_id: str | None = Field(default=None, alias="sectionId")
    knowledge_point_ids: list[str] | None = Field(default=None, alias="knowledgePointIds")
    difficulty: str = "medium"
    questions: list[dict[str, Any]] | None = None
    source: str = "llm_generated"


class QuizResponse(BaseModel):
    """Serialised QuizModel returned to the frontend."""

    id: str = ""
    title: str = ""
    session_id: str = Field(default="", alias="sessionId")
    scope_type: str = Field(default="knowledge_point", alias="scopeType")
    scope_id: str | None = Field(default=None, alias="scopeId")
    path_id: str | None = Field(default=None, alias="pathId")
    stage_id: str | None = Field(default=None, alias="stageId")
    chapter_id: str | None = Field(default=None, alias="chapterId")
    section_id: str | None = Field(default=None, alias="sectionId")
    knowledge_point_ids: list | None = Field(default=None, alias="knowledgePointIds")
    difficulty: str = "medium"
    question_count: int = Field(default=0, alias="questionCount")
    questions: list | None = None
    source: str = "llm_generated"
    archive_policy: str = Field(default="none", alias="archivePolicy")
    created_at: str | None = Field(default=None, alias="createdAt")


# ═══════════════════════════════════════════════════════════════════════
# Exam Set
# ═══════════════════════════════════════════════════════════════════════


class ExamSetCreate(BaseModel):
    """Payload for creating a new exam set."""

    session_id: str = Field(default="", alias="sessionId")
    title: str = ""
    scope_type: str = Field(default="chapter", alias="scopeType")
    scope_id: str | None = Field(default=None, alias="scopeId")
    path_id: str | None = Field(default=None, alias="pathId")
    stage_id: str | None = Field(default=None, alias="stageId")
    chapter_id: str | None = Field(default=None, alias="chapterId")
    knowledge_point_ids: list[str] | None = Field(default=None, alias="knowledgePointIds")
    difficulty: str = "medium"
    difficulty_distribution: dict[str, int] | None = Field(default=None, alias="difficultyDistribution")
    question_count: int = Field(default=0, alias="questionCount")
    questions: list[dict[str, Any]] | None = None
    estimated_minutes: int = Field(default=30, alias="estimatedMinutes")
    total_score: int = Field(default=100, alias="totalScore")
    source: str = "llm_generated"
    archive_policy: str = Field(default="archive", alias="archivePolicy")


class ExamSetUpdate(BaseModel):
    """Payload for updating an exam set (partial update)."""

    title: str | None = None
    status: str | None = None
    question_count: int | None = Field(default=None, alias="questionCount")
    questions: list[dict[str, Any]] | None = None
    difficulty_distribution: dict[str, int] | None = Field(default=None, alias="difficultyDistribution")
    estimated_minutes: int | None = Field(default=None, alias="estimatedMinutes")
    total_score: int | None = Field(default=None, alias="totalScore")


class ExamSetResponse(BaseModel):
    """Serialised ExamSetModel returned to the frontend."""

    id: str = ""
    title: str = ""
    session_id: str = Field(default="", alias="sessionId")
    scope_type: str = Field(default="chapter", alias="scopeType")
    scope_id: str | None = Field(default=None, alias="scopeId")
    path_id: str | None = Field(default=None, alias="pathId")
    stage_id: str | None = Field(default=None, alias="stageId")
    chapter_id: str | None = Field(default=None, alias="chapterId")
    knowledge_point_ids: list | None = Field(default=None, alias="knowledgePointIds")
    difficulty: str = "medium"
    difficulty_distribution: dict | None = Field(default=None, alias="difficultyDistribution")
    question_count: int = Field(default=0, alias="questionCount")
    questions: list | None = None
    estimated_minutes: int = Field(default=30, alias="estimatedMinutes")
    total_score: int = Field(default=100, alias="totalScore")
    status: str = "not_started"
    source: str = "llm_generated"
    archive_policy: str = Field(default="archive", alias="archivePolicy")
    created_at: str | None = Field(default=None, alias="createdAt")
    updated_at: str | None = Field(default=None, alias="updatedAt")


# ═══════════════════════════════════════════════════════════════════════
# Attempt
# ═══════════════════════════════════════════════════════════════════════


class AttemptCreate(BaseModel):
    """Payload for starting a new attempt on a quiz or exam set."""

    session_id: str = Field(default="", alias="sessionId")
    quiz_id: str | None = Field(default=None, alias="quizId")
    exam_set_id: str | None = Field(default=None, alias="examSetId")
    max_score: int = Field(default=100, alias="maxScore")


class AttemptUpdate(BaseModel):
    """Payload for updating an in-progress attempt (e.g. saving progress)."""

    answers: list[dict[str, Any]] | None = None
    status: str | None = None


class AttemptSubmit(BaseModel):
    """Payload for submitting an attempt for grading."""

    answers: list[dict[str, Any]] = Field(default_factory=list)
    total_score: int | None = Field(default=None, alias="totalScore")


class AttemptResponse(BaseModel):
    """Serialised AttemptModel returned to the frontend."""

    id: int = 0
    attempt_id: str = Field(default="", alias="attemptId")
    session_id: str = Field(default="", alias="sessionId")
    quiz_id: str | None = Field(default=None, alias="quizId")
    exam_set_id: str | None = Field(default=None, alias="examSetId")
    learner_id: str | None = Field(default=None, alias="learnerId")
    answers: list | None = None
    total_score: int | None = Field(default=None, alias="totalScore")
    max_score: int = Field(default=100, alias="maxScore")
    status: str = "in_progress"
    started_at: str | None = Field(default=None, alias="startedAt")
    submitted_at: str | None = Field(default=None, alias="submittedAt")
    created_at: str | None = Field(default=None, alias="createdAt")
